Implement stream_vt100_support user function

mlocati opened the pull request:
This is a followup of https://bugs.php.net/bug.php?id=72768

TL;DR Console apps of newer Windows 10 versions support colors, but we need to manually set the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` console flag.

Here's my proposal of a new userland function: `bool stream_vt100_support(resource stream[, bool enable])`

Where:
- `stream` should be `STDOUT` or`STDERR`
- if `enable` missing: the function returns true if the specified stream supports colors
- if `enable` is true: the function tries to enable color support for the stream (return true on success)
- if `enable` is false: the function tries to disable color support for the stream (return true on success)

To be done:
- [X] currently the function accepts a string (eg `php://stdout`): it should instead accept a stream resource
- [X] tests

mlocati commented:
Any hint about how to retrieve the standard Windows handles (`STD_OUTPUT_HANDLE`/`STD_ERROR_HANDLE`) starting from a `php_stream`?

mlocati commented:
> - [x] currently the function accepts a string (eg php://stdout): it should instead accept a stream resource

Done in https://github.com/php/php-src/pull/2103/commits/bd93781159e977a4d5a8d5620838fffc061192a1  :wink:

mlocati commented:
> - [x] tests

Done in https://github.com/php/php-src/pull/2103/commits/d9b1c99a0233f218bf0137e9fad2e0923b06836a  :wink:

KalleZ commented:
Hi

Wouldn't it be easier if this was added as a context option instead?

mlocati commented:
> Wouldn't it be easier if this was added as a context option instead?

What exactly do you mean? An example?

mlocati commented:
Sample code:
```php
<?php
$sampleText = "\033[101;93m Yellow text on red background \033[0m\n";
echo "Enabled: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";
echo $sampleText;
echo "Enable: ", stream_vt100_support(STDOUT, true) ? "ok" : "failed", "\n";
echo "Enabled: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";
echo $sampleText;
echo "Disable: ", stream_vt100_support(STDOUT, false) ? "ok" : "failed", "\n";
echo "Enabled: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";
echo $sampleText;
```

Output:
![image](https://cloud.githubusercontent.com/assets/928116/18092089/3cdb56a6-6ecb-11e6-8a88-b4686d619c25.png)

mlocati commented:
**Remark 1**: since the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` flag is associated to the whole console, if both STDOUT and STDERR are bound to the same console (ie: almost always if there's no redirection), the following code
```php
<?php
echo "Enabled on STDOUT: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";

echo "Enable on STDERR : ", stream_vt100_support(STDERR, true) ? "ok" : "failed", "\n";
echo "Enabled on STDOUT: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";

echo "Disable on STDERR: ", stream_vt100_support(STDERR, false) ? "ok" : "failed", "\n";
echo "Enabled on STDOUT: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";
```
ouputs
```
Enabled on STDOUT: no
Enable on STDERR : ok
Enabled on STDOUT: yes
Disable on STDERR: ok
Enabled on STDOUT: no
```

**Remark 2**: this function checks if the specified stream is redirected.
So, let's assume we have a `test.php` file like this:
```php
<?php
echo "Enable on STDOUT: ", stream_vt100_support(STDOUT, true) ? "ok" : "failed", "\n";
echo "Enable on STDERR: ", stream_vt100_support(STDERR, true) ? "ok" : "failed", "\n";
echo "\n";
echo "Enabled on STDOUT: ", stream_vt100_support(STDOUT) ? "yes" : "no", "\n";
echo "Enabled on STDERR: ", stream_vt100_support(STDERR) ? "yes" : "no", "\n";
```
- running `php test.php 2>out.txt` results in  
```
Enable on STDOUT: ok
Enable on STDERR: failed

Enabled on STDOUT: yes
Enabled on STDERR: no
```
- running `php test.php >out.txt`, the contents of `out.txt` is  
```
Enable on STDOUT: failed
Enable on STDERR: ok

Enabled on STDOUT: no
Enabled on STDERR: yes
```
- running `php test.php >out.txt 2>&1`, the contents of `out.txt` is  
```
Enable on STDOUT: failed
Enable on STDERR: failed

Enabled on STDOUT: no
Enabled on STDERR: no
```

weltling commented on run-tests.php:
```diff
@@ -1062,7 +1062,7 @@ function error_report($testname, $logname, $tested)
 }
 
-function system_with_timeout($commandline, $env = null, $stdin = null)
+function system_with_timeout($commandline, $env = null, $stdin = null, $captureStdOut = true, $captureStdErr = true)
```
It's a nice approach, but i'd rather avoid adding the new section to run-tests.php. It is not a general case and probably only useful for this concrete case, so IMO a new section is not justified.

What could be done instead, were doing an extra PHP run inside the test. Fe we do this for the CLI server, you can cehck sapi/cli/tests/php_cli_server.inc for a simple example framework. It is also better to be close to the real word, when testing redirection. Like

`php.exe -r "if (stream_vt100_support(STDOUT, true)) echo \"\033[101;93msupported\033[0m\n\"; else echo 'not supported'; " > tmp`

should should out "not supported" into a file on both linux and windows. Or i don't get this part right?

Thanks.

weltling commented:
@mlocati for the stream option, it could be similar like I've mentioned in the ticket, like

`stream_context_set_option($fd, "stdio", "enable_vt100", true);`

It were same as using the function. As for me, the stream option is not absolutely necessary in the current implementation. It would become necessary, if we'd go for an automation to strip the ANSI controls.

 Thanks.

johnstevenson commented:
@mlocati Great work in trying to get this incorporated (I have been following this via Symfony and Composer issues).

However, your implementation is incorrect in testing if the handle is redirected, rather than testing if it is a terminal (which is what `isatty` does on posix systems). Your function will only return true on a file, which means that when you give a pipe handle to `GetConsoleMode` it will fail (as it will on a file).

Microsoft have updated their  [vt100 documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/mt638032(v=vs.85).aspx) and state that:

> SetConsoleMode returning with STATUS_INVALID_PARAMETER is the current mechanism to determine when running on a down-level system

so there is no need for any version checking, which will make the code simpler. Likewise, you can use GetConsoleMode as the equivalent of isatty, as mentioned above (which is what [Docker](https://github.com/docker/docker/blob/master/pkg/term/windows/console.go) does).

mlocati commented:
> This should be zend_long64, at least for PHP 7.x series, see Zend/zend_types.h for more global types :)

Did you mean `zend_long`, right? I can't find any definition of `zend_long64`...

mlocati commented:
@johnstevenson Thanks! I'd keep a "is this stream a valid console" function, so that we can update it in case the current implementation needs future adjustments (btw, I replaced  `GetFinalPathNameByHandleW` with `GetConsole` as you suggested: it now works with pipes too, thanks!)

mlocati commented:
@weltling So, instead of
```c
#ifdef _WIN64
	(zend_long or long long) fileno;
#else
	int fileno;
#endif
```
I should write
```c
	intptr_t fileno;
```

?

weltling commented:
@mlocati yeah, that's fine. The (void *) would be exact same size platform dependent.

Thanks.

mlocati commented:
@weltling I used `zend_long`...

About `stream_vt100_support()` vs `stream_context_set_option()`: let's assume this PHP code:
```php
$ctx1 = stream_context_create();
stream_context_set_option($ctx1, 'stdio', 'enable_vt100', true);

$ctx2 = stream_context_create();
stream_context_set_option($ctx2, 'stdio', 'enable_vt100', false);
```

The problem here is that one may think that the two contexts are isolated, but they are not...

weltling commented:
@mlocati you're right. I was just testing and looking yet more over the patch. Please consider a snippet like

```
$fd = fopen("php://stdout", "w");
if (stream_vt100_support($fd, true)) {
    fwrite($fd, "\033[101;93msupported\033[0m\n"); 
} else {
    fwrite($fd, 'not supported');
}
```

In the current implementation it won't work, as the checks are focused on the original stdio handles. The `$fd` stream is however still associated with some valid console. IMO we should consider this as a bug. So far I see, the checks whether the underlying stream fd is the exact std handle is not necessary. The fileno should be simply passed to `php_win32_console*()` functions, and then _get_osfhandle should be used. The Windows console API should handle it correctly then.

Any php stream may point to a valid console, not just STDIO. I guess that set/get console mode will anyway apply to the given console object disregarding the fd used. Full isolation seems not to be possible therefore, as that's the way how the low level API works. But on the PHP side it should be possible to operate on `$fd` same way it's possible for STDIO directly. That will allow then to append filters to any vt100 enabled/disabled stream. 

Thanks

mlocati commented:
Thank you @weltling! `_get_osfhandle` does work really great (see https://github.com/php/php-src/pull/2103/commits/99bd01e53d780ccc4b24c7e9c744878adee0eeb4)!
I also added support for the stream names: see https://github.com/php/php-src/pull/2103/commits/7b155f48447ed9b07e5a798b13326a6533e5bd77

If everything is ok, I can start working on tests

weltling commented:
@mlocati no more remarks from my side. If there are no other comments, probably you should move forward with test updates. I'll do more thorough testing when the tests are ready as well. AFM it's already a self enclosed change, further tweaks can be done in a follow up PR.

Thanks.

mlocati commented:
About tests: I did a quick inspection and I don't think that using the php executable as a server like in `sapi/cli/tests/php_cli_server.inc` is the way to go: it seems that stdout/stderr are always redirected.

I'm going to replicate the approach of the `system_with_timeout` function of `run-tests.php`, but targeted to this specific subject (redirection of stdout/stderr).

weltling commented:
Yeah, I didn't mean a HTTP server is needed. Just a couple of functions that can be reused in multiple tests.

Thanks.

johnstevenson commented:
@mlocati You no longer need `php_win32_console_os_supports_vt100()` because everything is handled by the GetConsoleMode and SetConsoleMode API functions. In fact it is recommended as the:

> current mechanism to determine when running on a down-level system

and they don't mention anything about checking version numbers in the docs (so there is no guarantee that this method will always report correctly).

mlocati commented:
@johnstevenson Yes, you're right, I forgot about it: thanks!
Ref: https://github.com/php/php-src/pull/2103/commits/dd858250b794be84be7ba7721503d851dc6b3bcf

johnstevenson commented:
@mlocati Cool. Sorry, but I've got a few more comments. Can you confirm if you differentiate between input and output streams (it doesn't look like you do)?

mlocati commented:
> Can you confirm if you differentiate between input and output streams (it doesn't look like you do)?

@johnstevenson I do, but as long as the stdout and stderr streams are attached to the same console (99.99999% of cases), the system does not differentiate them: it's a Windows console level settings

johnstevenson commented:
@mlocati Using stream names is incomplete because you have not included `php://fd/?` values.

However, I would question whether the function signature (_resource or url_) is correct here, because all the _stream__ functions that perform work on a stream, only require a resource (the exception is `stream_is_local`, which takes a _resource or url_, and makes sense because it is intended to work on all stream types).

But that is not the case here because `stream_vt100_support` only deals with STDIO streams. And anyway, the user can pass in one of the `STDIN STDOUT STDERR` stream constants to get the same result. Perhaps @weltling could advise?



`

johnstevenson commented:
@mlocati Sorry, but I don't understand what you mean by:

> the system does not differentiate them: it's a Windows console level settings

I was asking about input and output streams (which are entirely different beasts!)

mlocati commented:
> I was asking about input and output streams (which are entirely different beasts!)

Indeed they are. But let's assume I call something like this (pseudocode)
```c
SetConsoleMode(STDOUT, ENABLE_VIRTUAL_TERMINAL_PROCESSING);
```
the Windows system applies the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` to the rendering layer of the console.
So, even when you write to stderr, Windows sees the flag `ENABLE_VIRTUAL_TERMINAL_PROCESSING`.

Am I missing something?

mlocati commented:
@weltling I tried to adopt a different approach for tests (see my [vt100-windows-newtests](https://github.com/mlocati/php-src/commits/vt100-windows-newtests) branch).

Here's what I did:
- in [e7d1cf8](https://github.com/mlocati/php-src/commit/e7d1cf893fed8039dd8af3e9186e770c2ea9797d) I reverted the edits I did to the `run-tests.php` file.
- in [11d7e8f](https://github.com/mlocati/php-src/commit/11d7e8f5536d2e02ca86f85124a58259006b5fa8) I introduced a new class that runs PHP code with the possibility to specify redirecting/piping/do-nothing to stdout/stderr.

BTW, since tests are executed by the standard  `run-tests.php` file, and since it run the tests piping the tests output, the new class of mine can't work as expected, since stdout/stderr are always piped...

Any ideas about how I could solve this?

johnstevenson commented:
@mlocati I was asking about differentiating between input (`STDIN`) and output (`STDOUT STDERR`).

However, regarding your comment about output:
> the Windows system applies the ENABLE_VIRTUAL_TERMINAL_PROCESSING to the rendering layer of the console.

I don't think you can guarantee this behavior (for example this doesn't happen in a powershell console).

mlocati commented:
Whoops, sorry :wink:
If you try to set the flag for `STDIN` it will fail (as expected). Furthermore, if you set the flag for `STDOUT` or `STDERR`, `STDIN` won't be affected (as expected).

mlocati commented:
For instance, if you run this code (tested both in cmd and powershell):
```php
stream_vt100_support(STDOUT, false);
echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'Enable for STDOUT: ', var_dump(stream_vt100_support(STDOUT, true));
echo 'Enable for STDIN: ', var_dump(stream_vt100_support(STDIN, true));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
```

here's the outout:
```
STDIN state: bool(false)
STDOUT state: bool(false)
Enable for STDOUT: bool(true)
Enable for STDIN: bool(false)
STDOUT state: bool(true)
STDIN state: bool(false)
```

mlocati commented:
No, sorry, I was wrong (I launched this code inside a class that redirected the stdin).

The correct output of
```php
stream_vt100_support(STDIN, false);
stream_vt100_support(STDOUT, false);
stream_vt100_support(STDERR, false);

echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'STDERR state: ', var_dump(stream_vt100_support(STDERR));

echo 'Enable for STDIN: ', var_dump(stream_vt100_support(STDIN, true));
echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'STDERR state: ', var_dump(stream_vt100_support(STDERR));

echo 'Disable for STDIN: ', var_dump(stream_vt100_support(STDIN, false));
echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'STDERR state: ', var_dump(stream_vt100_support(STDERR));

echo 'Enable for STDOUT: ', var_dump(stream_vt100_support(STDOUT, true));
echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'STDERR state: ', var_dump(stream_vt100_support(STDERR));

echo 'Enable for STDERR: ', var_dump(stream_vt100_support(STDERR, true));
echo 'STDIN state: ', var_dump(stream_vt100_support(STDIN));
echo 'STDOUT state: ', var_dump(stream_vt100_support(STDOUT));
echo 'STDERR state: ', var_dump(stream_vt100_support(STDERR));
```

is

```
STDIN state: bool(false)
STDOUT state: bool(false)
STDERR state: bool(false)
Enable for STDIN: bool(true)
STDIN state: bool(true)
STDOUT state: bool(false)
STDERR state: bool(false)
Disable for STDIN: bool(true)
STDIN state: bool(false)
STDOUT state: bool(false)
STDERR state: bool(false)
Enable for STDOUT: bool(true)
STDIN state: bool(false)
STDOUT state: bool(true)
STDERR state: bool(true)
Enable for STDERR: bool(true)
STDIN state: bool(false)
STDOUT state: bool(true)
STDERR state: bool(true)
```

So it seems that Windows supports associating the flag to `STDIN`, and it's differentiated to the `STDOUT`/`STDERR` pair.

BTW, I don't think I can do something about it: it's all handled by Windows...

johnstevenson commented:
> BTW, I don't think I can do something about it: it's all handled by Windows...

Sorry but Windows has nothing to do with it. Try running this command:

`php -r "var_dump(stream_vt100_support(STDIN));"`

and the result will be an unexpected: `bool(true)` 

This is because you are checking if the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` flag (0x0004) is set on an **input** handle. And indeed it is, because the console input buffer uses a different set of flags and in this case the value 0x0004 refers to `ENABLE_ECHO_INPUT`. Switching this off for the user would not be helpful!

So you must not process STDIN. I think that the best way of doing this is to check the stream->orig_path, which is what is used to provide the uri value in stream_get_meta_data. Something like (untested):

```c
if (stream->orig_path) {
    if (!strcmp(stream->orig_path, "php://stdin") || !strcmp(stream->orig_path, "php://fd/0")) {
        RETURN_FALSE;
    }
}
```
Perhaps @weltling  could confirm if this is the best method (and when the orig_path might be null). 

Note that you cannot reliably check that `stream->mode` is `"rb"` and `stream->ops->label` is `"STDIO"` because you could open a stdin stream like this: `fopen('php://stdin', 'wb')`.

weltling commented:
@mlocati pah, i see the issue with tests. My bad not understanding it at the start :( Look your first test approach was correct. Usually we avoid editing run-tests.php, in many cases things can be solved locally in a test itself like mentioned with example for cli server. But obviously, any test is being executed from within run-tests.php which is the top level script. So there's no way to tell run-tests.php to avoid piping except introducing a new section. Thus, adding a new section is justified for this reason. Alternative is to leave some parts untested, which is not really good. Bad was it to cause your additional work :(

@johnstevenson with `php://fd/?` - that's a point, indeed. It might be tricky, as a URL like `php://fd/?` might be even not have been opened yet. Opening a stream internally just for a check is senseless in that case. Fe `$fd0 = fopen("php://stderr", "w"); $fd0 = fopen("php://stderr", "w");` are indeed two different descriptors. Even we know it's the same console, probably looks a bit confusing in PHP code. I'd suggest to only support an opened resource, as for strings it simply won't be substantial per se. What do you think, @mlocati?

Thanks.

mlocati commented:
@johnstevenson You're right, of course. Sorry for my (multiple) mistakes.

What about something like this:
```c
if (GetNumberOfConsoleInputEvents(handle, &cNumberOfEvents)) {
	flag = ENABLE_VIRTUAL_TERMINAL_INPUT;
}
else {
	flag = ENABLE_VIRTUAL_TERMINAL_PROCESSING;
}
```

From some this seems to correctly differentiate STDIN vs STDOUT/STDERR...

@johnstevenson + @weltling yes, removing string arguments is fine. IMHO it's already quite counterintuitive that stdout and stderr share the same flags :wink:

mlocati commented:
@weltling No problems about the tests work. I think we'd need also to add another flag to the new `CAPTURE_STDIO` to tell `system_with_timeout` to not pipe `STDIN`, so that we can add tests for it too...
Would it be ok?

johnstevenson commented:
@mlocati No worries. It is not easy getting your head around all this! However you don't want to set `ENABLE_VIRTUAL_TERMINAL_INPUT` on the input because it essentially puts the console into raw mode and makes it unusable (unless you are an "application that communicates exclusively via virtual terminal sequences", to quote the docs).

Also there is a bug (or a feature!) that when you try and set this value and it fails (because it is unsupported) the mode you tried to set will be returned in the next in-process call to GetConsoleMode, so you have to explicitly unset it in another call to SetConsoleMode. You can see [Docker](https://github.com/docker/docker/blob/master/pkg/term/term_windows.go#L49) do this.

`GetNumberOfConsoleInputEvents` works well (at least on Windows 10) and only succeeds on a valid console input handle.

However, this is all getting very unwieldy and complicated when all libraries want to know is whether the stream is a terminal and if it supports some form of vt100/ANSI output.

mlocati commented:
@johnstevenson Thanks for the hint, I didn't know it: see https://github.com/php/php-src/pull/2103/commits/157f8e86bd85b2415d1035e6dce62fdb638126fe (a note in the msdn manual would have helped...)

About dropping support to `STDIN`... I don't know... should we inhibit it (assuming that some unskilled php developer will have problems), or should we grant freedom of usage?

mlocati commented:
> However, this is all getting very unwieldy and complicated when all libraries want to know is whether the stream is a terminal and if it supports some form of vt100/ANSI output.

Well, all they have to to is:
```php
// Initialization
stream_vt100_support(STDOUT, (bool) $consoleShouldSupportColor));
// Usage
if (stream_vt100_support(STDOUT)) {
   //echo with ansi control codes
} else {
   //echo without ansi control codes
}
```
Having/not having support to STDIN is not influential for 99.999% of cases imho...

johnstevenson commented:
Actually the php code will be more complicated because we'd have to use `function_exists` and on Windows we don't know why it failed. Was it because the stream is not a terminal, or because the version of Windows doesn't support vt100 output (in which case then you need to do the environment variable checks for ConEmu, Ansicon and xterm)? On non-Windows platforms it is just another `posix_isatty`.

I'd prefer to see vt100 switched on if available at cli start-up (it will be anyway if you use Powershell) with no way of turning it off (afterall, why would you do that, anyway?).

weltling commented:
@mlocati yep, as we see now an additional section is unavoidable for tests, so should be done.

IMHO with `GetNumberOfConsoleInputEvents` and checking `fileno != 0` the most of rough mistakes would be catched. If there's a solution, probably should be used.

Thanks.

mlocati commented:
@weltling what do you think about dropping support to `STDIN` (I'm neutral about that) and trying to enable vt100 support at startup (+1 for me)?

johnstevenson commented:
FYI re enabling vt100 support at start up, here is the situation regarding some windows implementations:

- when vt100 support was first introduced, cmd.exe mainly enabled it on start-up (depending where you had started it from) and Powershell didn't.
- since the anniversary update cmd.exe does not enable it automatically, while Powershell now does (on both stdout and stderr if they are a terminal), and provides a new function to determine this `$host.UI.SupportsVirtualTerminal`
- if you echo vt100 output in a batch file, cmd.exe renders it correctly (even though cmd.exe has not enabled vt100 output)
- Docker and node.js both enable vt100 on start-up. These programs are slightly different because they already provided vt100 output (by manipulating the console output attributes etc) but now they revert to the native Windows implementation if it is available. They provide functions for determining if the output stream is a terminal, so the user can decide whether to include vt100 escape sequences.

Hope this helps.

weltling commented:
PHP strives to be cross platform. But the first policy is still the backward compatibility. AFM, the automatic enablement is not something absolutely required today. It was never enabled before, it is still a new shiny thing. If it's not enabled - no existing PHP app will be broken. Once we have the feature, this can be modified in later releases or even partially backported if we see it is stable enough. also, there's nothing wrong to give some control instruments. It's a strange wild world, so are the use cases. And it'd anyway need a fair bit on documentation for the user space :)

In the initial discussion in the PHP ticket 72768 , it was proposed to simply enable vt100. Many other questions was cleared,  but the main one still is about BC concerns and plain cmd.exe. Back then, the API docs was a bit different, that's why the patch in the early was version based. Still enable by default could be ok, if done carefully, tested down to win7 and shows no breakage. @mlocati, if you want to dire for such a solution, I can support you as well. Though, it's more testing work :) But IMO there's no reason to abandon the function which will give users a choice.

Thanks for the research, @johnstevenson. From your other comment about stream_vt100_support function - it's thought to be dedicated for the exact VT purpose. You're right about non Windows, as there it's just a redundant isatty(), but on Windows it's slightly more stuff. The fun thing i see now, posix_isatty and some other functions from ext/posix could be ported to Windows with more ore less effort. And they're indeed missing there. Well, that's not the goal here, but maybe it'd make sense to just reduce the actual PR to Windows only? @mlocati? Fe the naming scheme were sapi_windows_vt100_support(), sapi_windows_isatty(), etc. (called similar to functions that are already in 7.1 like sapi_windows_cp_conv). Besides exposing more of the API, it would actually actually solve the situation mentioned by @johnstevenson and give a user more power for other related things.

@mlocati with STDIN, as mentioned in https://github.com/php/php-src/pull/2103#issuecomment-252034635 - if we can prevent rough mistakes, imho it's worth it.

Thanks.

mlocati commented:
No problems for the user docs: I can write it of course.

About tests on other/older systems: no problems here too.

About STDIN: I'll drop support to it.

And finally about Windows only functions: yes, for posix systems stream_vt100_support is almost the same as posix_isatty. BTW IMHO that does not mean that we should keep Windows only and posix only functions: it would be much better to call the same functions in every environment without checking PHP_OS or function_exists...
So, instead of having duplicated functions I'd update existing ones to add support to Windows  (if compatible and if function names are not OS specific, like posix_...): in the other cases I'd add new functions accessible by every OS (like this stream_vt100_support that in practice make posix_isatty deprecated).

PS: what does AFM stand for?

(Sorry for the brevity: I'm on a mobile phone)

mlocati commented:
I performed some tests on Windows 7 32 bit, and everything's working fine.

Just a side note... I run the following code on Win7:
```c
DWORD mode;

printf("Getting initial value... ");
if (!GetConsoleMode(handle, &mode)) {
	printf("FAILED!\n");
} else {
	printf("%#010x\n", mode);
	mode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
	printf("Setting new mode: %#010x... ", mode);
	if (SetConsoleMode(handle, &mode)) {
		printf("ok\n");
	} else {
		printf("FAILED!\n");
	}
	printf("Getting final value... ");
	if (GetConsoleMode(handle, &mode)) {
		printf("%#010x\n", mode);
	} else {
		printf("FAILED!\n");
	}
}
```
its output is:
```
Getting initial value... 0x00000003
Setting new mode: 0x00000007... FAILED!
Getting final value... 0x00000003
```

As you can see, if SetConsoleMode fails, the previous mode is preserved, so it seems that https://github.com/php/php-src/pull/2103/commits/157f8e86bd85b2415d1035e6dce62fdb638126fe is not necessary (but let's keep it anyway).

johnstevenson commented:
@mlocati There is no need for it anyway, because the behavior only happens when you call SetConsoleMode on the console's input buffer. I was actually referring to `ENABLE_VIRTUAL_TERMINAL_INPUT`. Sorry if that was not clear.

weltling commented:
@locati AFM - as for me.

Good point about the function names, too. When I try to map between Linux and Widows part

| Windows | Linux |
|-----------|------|
| php_win32_console_fileno_is_console | isatty |
| php_win32_console_fileno_has_vt100 | true in most case, otherwise unknown |
| php_win32_console_fileno_set_vt100 | enabled in most case, otherwise unknown |

My motivation to suggest the separation was as follows. Facts

. posix_isatty is only supposed to answer the question "is it a console?"
. posix_isatty accidentially answers "is VT supported?" on Linux, by luck it might be even VT100 if not a prehistoric terminal
. stream_vt100_support answers both questions at once, while simple "is it a console?" answer is currently not possible on Windows
. there's an additional operation to enable/disable VT100 Windows only
. stream_vt100_support($fd, true) is fake on Linux
. stream_vt100_support($fd, false) has no effect on Linux

As one sees, the current stream_vt100_support() doesn't make posix_isatty deprecated. It introduces a redundant functionality on Linux, while not answering "is output redirected?" on Windows. Sure, it wasn't the initial goal, it just came out in the discussion. To compare, there was other functions like getrusage() or ftok(), that was ported to windows with full compatibility, but the situation with vt100 is a bit more ambiguous.

Many POSIX APIs are supported on Windows, true. With posix_* functions the tricky part is that it's packed into ext/posix. While I'm sure ext/posix could be made partially available on Windows, a big part of it is not applicable. Maybe some could be emulated. posix_isatty is a simple thing, anyway. I just wanted to avoid a huge porting, where a simple way is possible. As mentioned, it doesn't belong to the actual PR directly. It is more about the global picture in PHP. Since this PR is on the related area, it should be taken into account. I'm going to check with ext/posix.

Thanks for you work!

johnstevenson commented:
@weltling I wonder if it would be worth splitting this into 2 simple functions:

- `stream_isatty()` 
- `sapi_windows_vt100_output()`

That way we have a cross platform `isatty` and a specific Windows function that returns (or sets) whether the console has native vt100 support. After that, it is up to developers to use them how they wish. For example:

```php
// Cross-platform isatty
function isTerminal($fd)
{
    if (function_exists('stream_isatty')) {
        return stream_isatty($fd);
    }

    if (DEFINED(PHP_WINDOWS_VERSION_MAJOR)) {
        $stat = fstat($fd);
        return 0020000 === ($stat['mode'] & 0170000);
    }

    return function_exists('posix_isatty') && posix_isatty($fd);
}

// Cross-platform color support - stdin is not checked here
function hasColorOutput($fd)
{
    if (!$result = isTerminal($fd)) {
        return false;
    } elseif (!DEFINED(PHP_WINDOWS_VERSION_MAJOR)) {
        return true;
    }

    if (function_exists(sapi_windows_vt100_output) && sapi_windows_vt100_output($fd)) {
        return true;
    }

    return false !== getenv('ANSICON')
        || 'ON' === getenv('ConEmuANSI')
        || 'xterm' === getenv('TERM');
}

```

mlocati commented:
While a decision about the future of this `stream_vt100_support` is taken, I added checks for all the 3 kinds of environment currently supported (so that we can consolidate the functions we currently have):
- `winok`: newer Windows versions (≥ 10.0.10586)
- `winko`: older Windows versions (< 10.0.10586)
- `posix`

I included in the PHP test suite [this test](https://github.com/mlocati/php-src/blob/7d348a04049f101719ee27597e12109cafdca63a/tests/output/vt100_support.php), that's executed by all the 3 environment, capturing STDIN and/or STDOUT and/or STDERR (if you have more test cases please let me know).

You can see the (successful) tests here: https://github.com/php/php-src/pull/2103/commits/7d348a04049f101719ee27597e12109cafdca63a

PS: I executed these tests on Win10 64 bit v10.0.14393 (for `winok`), Win7 32 bit (for `winko`) and Ubuntu Server  16.04.1 LTS (for `posix`)

mlocati commented:
So, since 

> Still enable by default could be ok, if done carefully, tested down to win7 and shows no breakage.

So, since:

1. the tests are now comprehensive and successful
2. the functions are harmless on older Windows versions
3. currently we have 2 different behaviours (`ENABLE_VIRTUAL_TERMINAL_PROCESSING` off in `cmd.exe` and on in PowerShell)

what about adding
```c
php_win32_console_fileno_set_vt100(STDOUT, TRUE);
php_win32_console_fileno_set_vt100(STDERR, TRUE);
```
at php cli startup?

weltling commented:
@mlocati it could be added in php_cli.c:main probably. I had no chance to test the latest code yet, but will do it. As well will ask a collegue for more extensive tests, but that might take some time. 

@mlocati not sure, why you say the decision is taken? It looked like we were discussing an improvement on this? In my opinion, the @johnstevenson's suggestion makes sense. stream_isatty() can be available even in a minimal build and crossplatform. And, instead of having stream_vt100_support doing void on Linux, it can be moved into a Windows only namespace. IMHO a good middle course variant, both flexible and giving a new feature for the user instead of keeping same function for the sake of it. That's an opportunity to improve in general and to add vt100 at the same time. What makes you to reject it? The actual low level code is already written, making two functions from one is not a big deal.

Thanks.

weltling commented on tests/output/vt100_support.php:
```diff
@@ -0,0 +1,65 @@
+<?php
```
The file should be .inc, not .php. Otherwise there might be issues when temporary files are cleaned up.

Thanks.

johnstevenson commented:
@weltling I read _While a decision about the future of this stream_vt100_support is taken_ as meaning that during the time this decision is being made, here are some tests anyway.

Regarding enabling vt100 at start up, I've posted a question on the Microsoft forum, in the hope that it might provide some guidance.

[https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/16547887-what-is-the-rationale-for-cmd-exe-rendering-vt100](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/16547887-what-is-the-rationale-for-cmd-exe-rendering-vt100)

If we do go down this route, then there is another function option:

```
bool stream_isatty(resource stream[, bool &vt100_output])
Determine if the stream is a tty and optionally supports vt100 escape sequences

Parameters
stream
    The stream
    
vt100_output
    If vt100_output is provided and the stream is either stdout or stderr, then
    this value is set to TRUE if vt100 escape sequences are supported or FALSE
    if they are not. On Windows this is dependent on native operating system
    support, on other platforms this value will always be TRUE. Note that this
    value is initially set to FALSE.
    
Return Values
    Returns TRUE if the stream is a tty or FALSE if it is not. FALSE will also be
    returned if vt100_output is provided and the stream is stdin.
```
Personally I am uncomfortable with the fact that calling it with a stdin stream and the vt100_output param will always return false, which makes it a less useful function than posix_isatty.

I much prefer separating the different platform behaviors using 2 functions and have included this here just for completeness.

Locati commented:
Offtopic: Somehow Github mixed up my account and Mr. Locatis. I keep
getting mails to this subject because he is mentioned.

Am 08.10.2016 12:21 schrieb "John Stevenson" <notifications@github.com>:

@weltling <https://github.com/weltling> I read *While a decision about the
future of this stream_vt100_support is taken* as meaning that during the
time this decision is being made, here are some tests anyway.

Regarding enabling vt100 at start up, I've posted a question on the
Microsoft forum, in the hope that it might provide some guidance.

https://wpdev.uservoice.com/forums/266908-command-prompt-
console-bash-on-ubuntu-on-windo/suggestions/16547887-
what-is-the-rationale-for-cmd-exe-rendering-vt100

If we do go down this route, then there is another function option:

bool stream_isatty(resource stream[, bool &vt100_output])
Determine if the stream is a tty and optionally supports vt100 escape sequences

Parameters
stream
    The stream

vt100_output
    If vt100_output is provided and the stream is either stdout or stderr, then
    this value is set to TRUE if vt100 escape sequences are supported or FALSE
    if they are not. On Windows this is dependent on native operating system
    support, on other platforms this value will always be TRUE. Note that this
    value is initially set to FALSE.

Return Values
    Returns TRUE if the stream is a tty or FALSE if it is not. FALSE
will also be
    returned if vt100_output is provided and the stream is stdin.

Personally I am uncomfortable with the fact that calling it with a stdin
stream and the vt100_output param will always return false, which makes it
a less useful function than posix_isatty.

I much prefer separating the different platform behaviors using 2 functions
and have included this here just for completeness.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/php/php-src/pull/2103#issuecomment-252416875>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AHKyuoCXmb3VpEVXwY3ILCXoLJWP6Rjqks5qx26ggaJpZM4JwRDI>
.

weltling commented:
@locati, it's my bad. I made a typo in one of the previous comments which led to your nick mentioning. Please just unfollow the thread, that'll be it.

Thanks.

weltling commented:
@johnstevenson I'd see two functions better as well. I'd see not much sense in overloading isatty. Declaring `&$enable` prevents usage like `stream_isatty($fd, true)`. Otherwise it might be ok, but actually isatty is quite a simple bit so it's an unnecessary overload. Also that's the only crossplatform piece in the story. The other one, maybe could be like sapi_windows_vt100 for shorter, or sapi_windows_stream_vt100 for more precise naming. From the today's angle, IMO having two function by your first suggestion is the most concise and semantically correct way to go.

Thanks.

mlocati commented:
> @mlocati it could be added in php_cli.c:main probably.

Done in https://github.com/php/php-src/pull/2103/commits/5d2ecf9979a997c8bb1a62b346e7b07ba38798e3

> @weltling I read *While a decision about the future of this stream_vt100_support is taken* as meaning that during the time this decision is being made, here are some tests anyway.

Yes, that's exactly what I meant, in particular because @weltling [wrote](https://github.com/php/php-src/pull/2103#issuecomment-252247784) *I'm going to check with ext/posix.*

> What makes you to reject it?

I'm sorry but clearly I didn't express myself well... Indeed I do really like the `stream_isatty()`/`sapi_windows_vt100_output()` solution, and if tomorrow it will rain I can replace `stream_vt100_support` with them :wink:

mlocati commented on tests/output/vt100_support_posix_err.phpt:
```diff
@@ -0,0 +1,140 @@
+STDERR
+--FILE--
+<?php
+require 'vt100_support.php';
```
Fixed in https://github.com/php/php-src/pull/2103/commits/bf60a7880407c92b1b0bace0d1d48b6b5e777b9a

weltling commented:
@mlocati pah! I obviously missed something in my english class. Gramatically "Is taken" and "is being taken" is a similar construct. I apparently took the former as passive past, not as passive present, and obviously misunderstood the context. Thanks for the explanation on your point!

Yeah, I'd be indeed examining ext/posix closer, if we couldn't reach the common opinion. Meanwhile @johnstevenson made the other suggestion, so I thought the topic were switched. It's doable to make ext/posix parts to work on Windows, however it's a fair bit of work. It were a fun task, to be mentioned. I already had several drafts to support more from POSIX in the C FS layer, while working on another patch. But there seems to be no much demand in the community, so it is not really relevant ATM.

Seems the concept is clear now. 
 * vt100 is enabled by default automatically, where it's supported
 * stream_isatty() for crossplatform terminal check
 * sapi_windows_vt100_support() to check/control vt100 functionality 

I'm in process of preparing some test VMs, so i'll be ready to run the tests with the final variant. Anyway, have a lot of sun tomorrow, still it's a sunday :-)

Thanks.

mlocati commented:
...rainy day :wink:

- replaced `stream_vt100_support` with `stream_isatty` + `sapi_windows_vt100_support` (https://github.com/php/php-src/pull/2103/commits/d15063c9e2bcc3d9c1dafe01d248e247900fbb82)
- added tests for `stream_isatty` (https://github.com/php/php-src/pull/2103/commits/65ad7796d824c2ea26316681867b157f6a9e597d)
- added tests for `sapi_windows_vt100_support` (https://github.com/php/php-src/pull/2103/commits/67e1eb7dad182bb5ca43d5b5d2764c4b652df47b)

weltling commented:
@mlocati nice! I did tests on win10, win8, server 2012 and Linux, both nmake test and some manual scenarios. Everything looks good so far. I waiting yet for more test results, that may take some time. 

To the doc part - at least some doc should be added into UPGRADING and UPGRADING.INTERNALS, respectively to what is relevant. Adding a patch to the manual directly were also great, if you had time.

Possible further directions could be checked
 - recognition of terminals with vt100, useful for pre win10 build 1607
 - stream filter to handle automatic control sequences stripping
 - possibility for a partial backport (not at once, but after some time)
 - integration with phpdbg
 
But that probably in separate follow ups. At this point, probably it's good to let the PR to ripen a bit, let everyone interested to check and test.

Thanks.

mlocati commented:
Currently we have this code:
```c
PHP_FUNCTION(stream_isatty)
{
#ifdef PHP_WIN32
	// Windows logic
#elif HAVE_POSIX
	// Posix logic
#else
	RETURN_FALSE;
#endif
}
```

What about changing the logic a bit?

```c
#if defined(PHP_WIN32) || HAVE_POSIX
PHP_FUNCTION(stream_isatty)
{
#ifdef PHP_WIN32
	// Windows logic
#else
	// Posix logic
#endif
}
#endif
```

I mean, I don't know if/when this may be applicable, but if the system is neither Windows nor Posix the function always returns false, but IMHO that's wrong (it's a case when we don't know if the stream is linked to a tty...).

mlocati commented:
Another alternative could be

```c
PHP_FUNCTION(stream_isatty)
{
#ifdef PHP_WIN32
	// Windows logic
#elif HAVE_POSIX
	// Posix logic
#else
	RETURN_NULL;
#endif
}
```

weltling commented:
@mlocati good catch, indeed! I'd be for keeping the function and returning NULL. In vast majority of cases POSIX is supported, so that would be a rare case anyway.

Thanks.

johnstevenson commented:
For those cases where POSIX is not supported, you could use fstat and mimic the `S_ISCHR` macro:

```c
PHP_FUNCTION(stream_isatty)
{
#ifdef PHP_WIN32
    // Windows logic
#elif HAVE_POSIX
    // Posix logic
#else
    zend_stat_t buf;
    if (zend_fstat(fileno, &buf) == 0) {
        if ((buf.st_mode & 0170000) == 0020000) {
            RETURN_TRUE;
        }
    }    
    RETURN_NULL;
#endif
}
```
I haven't tested this.

mlocati commented:
@weltling Done in https://github.com/php/php-src/pull/2103/commits/f9ff470b8c82fdcdbbae1de1356f7002c02f5016
@johnstevenson Yesh, why not... Done in https://github.com/php/php-src/pull/2103/commits/6df8c3c934e18ab57d30d0b1a6297d092e5ad9ba

@weltling + @johnstevenson I don't know if the PHP core team already considered it, btw what about integrating appveyor in this github repo?

mlocati commented:
> I don't know if the PHP core team already considered it, btw what about integrating appveyor in this github repo?

I mean, something like [this](https://github.com/mlocati/php-src/commit/f852c61d98bd7ae44605fa75de08088ce43b06b9) (you can find this commit in my [appveyor](https://github.com/mlocati/php-src/tree/appveyor) branch).

weltling commented:
@mlocati isn't appveyor a commercial tool? For php-src repo snaps are being produced on every push http://windows.php.net/snapshots/ . The test results are available as well https://qa.php.net/pftt.php . The test tool is called PFTT, which runs tests not only on CLI, but also Apache and other scenarios. However no github integration, yep.

Thanks.

mlocati commented:
Nope, appveyor is free for open-source projects... And here's a (still to be fixed) sample result: https://ci.appveyor.com/project/mlocati/php-src/build/1.0.14/job/9fn9945pe3q4apn4

johnstevenson commented on sapi/cli/php_cli.c:
```diff
@@ -1208,6 +1212,11 @@ int main(int argc, char *argv[])
 	 */
 	argv = save_ps_args(argc, argv);
 
+#ifdef PHP_WIN32
```
This also needs an `#ifndef PHP_CLI_WIN32_NO_CONSOLE` guard.

Maybe it is worth incorporating a few lines down in the `#ifdef PHP_WIN32` block that call setmode()? However to keep things consistent you would need to replace the `_fileno(std...)` calls with the `STD..._FILENO` defines.

mlocati replied:
> This also needs an #ifndef PHP_CLI_WIN32_NO_CONSOLE guard.

I'm don't know PHP internals that much, but a 3rd party app could attach a console to a php.exe compiled with PHP_CLI_WIN32_NO_CONSOLE?

mlocati replied:
> call setmode()?

Why?

johnstevenson replied:
Perhaps (why would you do that?), but I was basing this on the fact that the new codepage stuff uses this guard. Also, `_get_osfhandle` returns -2 if the handle is invalid (rather than -1).

johnstevenson replied:
Re setmode() - I meant in the block of code that calls setmode, just to keep the `PHP_WIN32` startup stuff together.

mlocati replied:
> Re setmode() - I meant in the block of code that calls setmode, just to keep the PHP_WIN32 startup stuff together.

They have different and unrelated purposes, I don't see why we have the necessity to keep them together.

weltling replied:
PHP_CLI_WIN32_NO_CONSOLE builds a binary without any interaction, which is meant to be run a a service for example. You can check sapi/cli/config.w32. Though i'm not sure how often this feature is actually used, I don't recall any related bug reports within last couple of years. Anyway  there is supposed to be no console interaction. Except maybe that the service is allowed to communicate with desktop, but even in that case - wouldn't the descriptors be redirected?

Thanks.

mlocati replied:
> even in that case, wouldn't the descriptors be redirected?

I guess so... https://github.com/php/php-src/pull/2103/commits/d1772e6c16d92c05048b53ac50ee97dfb77f3c9f

johnstevenson commented on ext/standard/streamsfuncs.c:
```diff
@@ -1635,6 +1636,121 @@ PHP_FUNCTION(stream_supports_lock)
+	php_stream *stream;
+	zend_long fileno;
+
+	int argc = ZEND_NUM_ARGS();
```
You only use this in the call to `zend_parse_parameters` now

mlocati replied:
:+1: https://github.com/php/php-src/pull/2103/commits/1081aa898fcca143f596aac4141ae0d306fc3837

johnstevenson commented on ext/standard/streamsfuncs.c:
```diff
@@ -1635,6 +1636,121 @@ PHP_FUNCTION(stream_supports_lock)
+	}
+
+#ifdef PHP_WIN32
+	/* Check if the Windows standard handle is redirected to file */
```
May be `/* Check if the file descriptor is a console */`

mlocati replied:
Much better: https://github.com/php/php-src/pull/2103/commits/e4aaa275295c65e23c5eb5170ec8341e75c7777a

johnstevenson commented on win32/console.c:
```diff
@@ -0,0 +1,98 @@
+	HANDLE handle = (HANDLE) _get_osfhandle(fileno);
+
+	if (handle != INVALID_HANDLE_VALUE) {
+		DWORD cNumberOfEvents;
```
Either `c_number_of_events` or perhaps just `events` to conform with the style guide.

mlocati replied:
Ok: https://github.com/php/php-src/pull/2103/commits/2d062f333bd7d6b61adbd5a78ee760429c748b85

johnstevenson commented on win32/console.c:
```diff
@@ -0,0 +1,98 @@
+						result = TRUE;
+					}
+					else {
+						// SetConsoleMode sets the flags even if it fails, so let's revert them back
```
You don't need to do this, it only affects an input console.

mlocati replied:
Ok: see https://github.com/php/php-src/pull/2103/commits/d70a7f288d6c144c4cf4c1b340272ab254ec9560

mlocati commented:
> _get_osfhandle returns -2 if the handle is invalid (rather than -1).

@johnstevenson I did some tests, and it always returned -1 (`INVALID_HANDLE_VALUE`). Do you have some use case when `_get_osfhandle` returns -2?

johnstevenson commented:
@mlocati If you debug php-win.exe and step through `php_win32_console_fileno_set_vt100` you can see the handle being returned as -2 (which makes `GetConsoleMode` fail anyway).

@weltling Blame me for suggesting the `PHP_CLI_WIN32_NO_CONSOLE` guard because I noticed that it was used around the `php_win32_cp_cli_do_setup()` code page stuff.

weltling commented:
@mlocati ok, in that case it might be interesting with appveyor. You rule actually :) There is a couple other things going on yet. Like the work on the new SDK tools, which interferes with related areas. All in all, IMO it's worth it to create a PR for the matter and discuss there.

@johnstevenson yeah, that's a similar case, as SetConsoleCP and co would fail as well.

Thanks.

weltling commented:
Merged with 33301d5bae4964f74fd1fc8c4fc485abfde0378e.

I've also added some basic doc to UPGRADING. @mlocati, if you've time for more doc - please either do a new PR, or gist the change and ping me, so it could be added there. Also, if you've time, you might check to integrate the enablement and handling into phpdbg.

Thanks!

mlocati commented:
Great!!
In these days my spare time is over, but I think I'll be more free in a week or so.

mlocati commented:
PS: what are the plans for these fubctions? I mean, for which php versions will them be available?

weltling commented:
@mlocati, it's merged into master, which is the future 7.2 nowadays. 7.1 was already freezed by the discussion time, anyway. And otherwise, it is fine to go through a good QA for the patches of such extent. The 7.2 cycle will start spring 2017, so there's enough time for possible fix ups and doc. I've already positive feedback from yet another tester, the PR is therefore fine for master. One thing standing out yet is the newly released Nano Server, as it has quite minimalist console capabilities. More or less, that's  the approx plan for this feature.

Thanks.

mlocati commented:
@weltling / @johnstevenson It seems that the `stream_isatty` function may have some issue when PHP is executed in Cygwin with Mintty:

```sh
$ php -r "var_export(stream_isatty(STDOUT));"
false
```

I'm not very familiar with the `_php_stream_cast` function...

weltling commented:
@mlocati so far i get with the git shell

```
weltling@w530-php-dev MINGW64 /c/php-sdk/phpmaster/vc15/x64/php-src (master)
$ x64/Release/php.exe -r "var_export(stream_isatty(STDOUT));"
true
```

it's based on some parts of Cygwin. Otherwise, no chance for me with pure Cygwin, as it's principally a workspace breaker when it comes to the native VS environment.

Is that a Cygwin build of PHP? The result above is with the native build. FYI, we don't officially support Cygwin and other similar build flavors.

Thanks.

mlocati commented:
> Is that a Cygwin build of PHP?

Nope: it's the php.exe version from https://windows.php.net/download/#php-7.2-ts-VC15-x86

mlocati commented:
```sh
$ uname -a
CYGWIN_NT-10.0 MYPC 2.10.0(0.325/5/3) 2018-02-02 15:16 x86_64 Cygwin

$ which php
/cygdrive/c/Dev/PHP/php

$ php -v
PHP 7.2.4 (cli) (built: Mar 28 2018 04:46:46) ( ZTS MSVC15 (Visual C++ 2017) x86 )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Xdebug v2.6.0, Copyright (c) 2002-2018, by Derick Rethans
    with Zend OPcache v7.2.4, Copyright (c) 1999-2018, by Zend Technologies

$ php -r "var_export(stream_isatty(STDOUT));"
false
```

johnstevenson commented:
Cygwin, Msys2 etc emulate pseudo terminals via named pipes, so this is correct (if unhelpful).

mlocati commented:
> Cygwin, Msys2 etc emulate pseudo terminals via named pipes, so this is correct (if unhelpful).

I've noticed the same problem in WLS. By the way, quite interestingly, with Git Bash this command
```sh
php.exe -r "var_export([stream_isatty(STDOUT), sapi_windows_vt100_support(STDOUT)]);"
```
outputs
```
array (
  0 => true,
  1 => true,
)
```

iworker commented:
This function (stream_isatty) is working **very** strange with Docker: when I'm running `make test` using specific RUN command:
```
RUN cd php-src-php-7.2.5 && TEST_PHP_ARGS="--show-out" TEST_PHP_DETAILED=1 TESTS=tests/output make test
```

from Dockerfile:

```
docker build -t app-php-72 .
```

I receive errors:
```
=====================================================================
Number of tests :   88                74
Tests skipped   :   14 ( 15.9%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    5 (  5.7%) (  6.8%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests passed    :   69 ( 78.4%) ( 93.2%)
---------------------------------------------------------------------
Time taken      :    2 seconds
=====================================================================

FAILED TEST SUMMARY
---------------------------------------------------------------------
Test stream_isatty with redirected STDERR [tests/output/stream_isatty_err.phpt]
Test stream_isatty with redirected STDIN/STDERR [tests/output/stream_isatty_in-err.phpt]
Test stream_isatty with redirected STDIN/STDOUT [tests/output/stream_isatty_in-out.phpt]
Test stream_isatty with redirected STDOUT/STDERR [tests/output/stream_isatty_out-err.phpt]
Test stream_isatty with redirected STDOUT [tests/output/stream_isatty_out.phpt]
```

When I'm running using `docker exec`, `php` and `run-tests.php`:
```
docker exec -it 0dd46d08dc29 sh -c "cd /php-src-php-7.2.5 && TEST_PHP_ARGS="--show-out" TEST_PHP_DETAILED=1 TESTS=tests/output make test"
```
I receive no errors:
```
=====================================================================
Number of tests :   88                74
Tests skipped   :   14 ( 15.9%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    0 (  0.0%) (  0.0%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests passed    :   74 ( 84.1%) (100.0%)
---------------------------------------------------------------------
Time taken      :    3 seconds
=====================================================================
```

Is somebody know what is the cause of such strange situation? :)
